### PR TITLE
Improve CJK character line breaking

### LINKED ISSUE #2963: Some languages don't text wrap because they don't utilise spaces in the language (eg chinese)
eg
image

We need to add some exceptions in place for some languages to allow them to wrap even if there are no spaces. This makes text unreadable when the text is resized to fit.

### PR DESCRIPTION
**Pull request by raytranuk**
- now deals with non-alphanumeric characters like punctuation.
- supports Korean and Japanese unicode ranges
- supports zero width space for Thai
- added some small kana and CJK punctuation characters that do not allow line break

Fixes #2963

I confirm I have signed the [Contributor License Agreement](https://docs.google.com/a/playcanvas.com/forms/d/1Ih69zQfJG-QDLIEpHr6CsaAs6fPORNOVnMv5nuo0cjk/viewform).

### CODE REVIEW

File: src/framework/components/element/text-element.js
Comment (slimbuck) [inline src/framework/components/element/text-element.js:907]: should this test be `(nextchar !== null) ? ...` ?

Comment (raytranuk) [reply src/framework/components/element/text-element.js:907]: yes - I'll change it

Comment (raytranuk) [reply src/framework/components/element/text-element.js]: done - thanks!

File: src/framework/components/element/text-element.js
Comment (slimbuck) [inline src/framework/components/element/text-element.js:907]:
```suggestion
                const isNextCJK = (nextchar !== null) && CJK_CHAR.test(nextchar);
```

and make the previous two variables const too!

File: src/framework/components/element/text-element.js
Comment (slimbuck) [inline src/framework/components/element/text-element.js:917]: I think the brackets should be like so:
```suggestion
                const isCJKBoundary = CJK_CHAR.test(char) && ((nextchar !== null) && (WORD_BOUNDARY_CHAR.test(nextchar) || ALPHANUMERIC_CHAR.test(nextchar)));
```

Comment (raytranuk) [reply src/framework/components/element/text-element.js]: thanks :)

File: src/framework/components/element/text-element.js
Comment (slimbuck) [inline src/framework/components/element/text-element.js:917]: How would you feel about this?

```suggestion
            // char is space, tab, or dash
            const isWordBoundary = () => {
                return WORD_BOUNDARY_CHAR.test(char);
            };

            const isValidNextChar = () => {
                return (nextchar !== null) && !NO_LINE_BREAK_CJK_CHAR.test(nextchar);
            };

            // char is a CJK character and next character is a CJK boundary
            const isNextCJKBoundary = () => {
                return CJK_CHAR.test(char) && (WORD_BOUNDARY_CHAR.test(nextchar) || ALPHANUMERIC_CHAR.test(nextchar));
            };

            // next character is a CJK character that can be a whole word
            const isNextCJKWholeWord = () => {
                return CJK_CHAR.test(nextchar);
            };

            if (isWordBoundary() || (isValidNextChar() && (isNextCJKBoundary() || isNextCJKWholeWord()))) {
```

Comment (slimbuck) [reply src/framework/components/element/text-element.js:917]: (NOTE: the proposed code above is untested!)

Comment (raytranuk) [reply src/framework/components/element/text-element.js:917]: I like it - it is more readable and I think the logic is equivalent - I'll commit it and just double check with my test case - thanks!

Comment (raytranuk) [reply src/framework/components/element/text-element.js]: I tested it and it seems to work - I also moved the function declarations out of the loop (as lint thinks its unsafe)

**Review by slimbuck — APPROVED** [review]
Nice job, approving! Pls wait for approval from someone else on the team before merging.

File: src/framework/components/element/text-element.js
Comment (willeastcott) [inline src/framework/components/element/text-element.js:934]: I find this `if` statement way easier to read now - nice. But personally, I would define the 4 functions at module scope near the top instead of inside _updateMeshes (where they'll be allocated on every call). They'd just need the char passing as an argument.

Comment (raytranuk) [reply src/framework/components/element/text-element.js]: makes sense - moved functions to module scope

**Review by willeastcott — APPROVED** [review]

**Pull request merged**
